### DISCUSSION TITLE
Query performance considerations when filtering data within a select statement in RavenDB

### DISCUSSION THREAD

**Discussion by montyclt** [ravendb/ravendb#16360] [Q&A] [answered]
Hello,

I have a RavenDB database that stores documents in the following format:

```json
{
    "ServiceName": "Mobile line 2 GB",
    "Customer": {...},
    "Addons": [
        {
            "Name": "Addon1",
            "Value": "Addon1Value",
            "IsActive": true,
        },
        {
            "Name": "Addon2",
            "Value": "Addon2Value",
            "IsActive": false,
        },
        {
            "Name": "Addon3",
            "Value": "Addon3Value",
            "IsActive": true,
        },
        {
            "Name": "Addon4",
            "Value": "Addon4Value",
            "IsActive": false
        },
    ]
}
```

I would like to query the document and filter the addons by their "IsActive" status. I have a couple of options:

- Store an additional property in the JSON document containing the filtered data
- Perform a projection using the .filter(function) method in RQL.
Here's an example RQL query:

```
from services as service
select {
    CustomerName: service.Customer.Name,
    CustomerId: service.Customer.Id,
    CustomerContract: service.Customer.Contract,
    ActiveAddons: service.Addons.filter(addon => addon.IsActive)
}
limit 0, 100
```

What are the performance considerations of using these queries compared to storing the filtered values directly? I'm using JavaScript to filter the data instead of the "where" operator in RQL, and I'm not sure if the "IsActive" field is involved in indexing.

As far as I know, RavenDB always uses indexes to resolve queries when the "where" operator is used in RQL. But what happens when filtering is done within the "select" statement?

Are there an important performance penalty, specially when filtering a large list of documents?

**Comment by ayende** [comment]
The query as stated will get *all* services. Is that expected, do you intend to have lot of those?

What is the _size_ of the `Addons` collection? Are you likely to have a _lot_ of those?

Basically, the scenario is that you have a tradeoff:

* Filtering on the server side (requires us to do JS projection & building of the output)
* Filtering on the client side, usually cheaper from the server perspective. But you need to read & parse the JSON then filter.

It's hard to give a concrete advice without the numbers, but in general, I wouldn't expect that to matter much.

**Reply by montyclt** [reply]
There are aprox. 8000 services today (adding 10-20 every month). To avoid overload de client memory, I query it on batches of 100. There are about 5-20 addons per service. Some services don't have any addon.

My question is about the impact doing this projection using JS. Maybe should I create a custom index to get the services with the filtered addons?

**Reply by ayende** [reply, marked as answer]
Are you needing *all* of them?
In that case, why not streaming the results?
Given the sizes involved, there is unlikely to be any meaningful difference in performance in either option
